pkg/controller/appinst_api.go

```diff
@@ -2372,6 +2370,54 @@ func (s *AppInstApi) RecordAppInstEvent(ctx context.Context, appInst *edgeproto.
 	services.events.AddMetric(&metric)
 }
 
+func (s *AppInstApi) updateURI(key *edgeproto.AppInstKey, cloudlet *edgeproto.Cloudlet, inCb edgeproto.AppInstApi_UpdateAppInstServer) (reterr error) {
+	ctx := inCb.Context()
+	cctx := DefCallContext()
+
+	log.SpanLog(ctx, log.DebugLevelApi, "updateURI", "appinst", key, "cloudlet", cloudlet)
+	streamCb, cb := s.all.streamObjApi.newStream(ctx, cctx, key.StreamKey(), inCb)
+
+	needUpdate := true
+	modRev, _ := s.sync.ApplySTMWaitRev(ctx, func(stm concurrency.STM) error {
+		appInst := edgeproto.AppInst{}
+		if !s.store.STMGet(stm, key, &appInst) {
+			// deleted in the meantime
+			log.SpanLog(ctx, log.DebugLevelApi, "Appinst deleted before DNS update", "appinst", key)
+			return nil
+		}
+		if appInst.Uri == "" {
+			needUpdate = false
+		}
+		// Store previous URI, so we can update this with CCRM
+		appInst.AddAnnotation(cloudcommon.AnnotationPreviousDNSName, appInst.Uri)
+		appInst.Uri = getAppInstFQDN(&appInst, cloudlet)
+		appInst.UpdatedAt = dme.TimeToTimestamp(time.Now())
+		appInst.State = edgeproto.TrackedState_UPDATE_REQUESTED
```

Comment:
Btw just realized, this needs to check the current appInst.State to make sure it's in Ready state, or if the user specified an appropriate crmIgnore state. You should be able to look at UpdateClusterInst for the logic.

Reply:
Hm, you are right - it will set the state to ready at the end of this. Also will need to check the state for clusterInsts.
For the simplicity I'd prefer to either skip the non-ready objects, or rather error our when we encounter the first one. Which way do you think is better?

With respect to crmIgnore - this api doesn't actually have it, since it takes a key for the cloudlet rather than the whole clouldet object.

Reply:
I think we skip the non-ready objects. I'm assuming we can rerun the command again and fix all the objects we missed the first time around.

Ok, I think we don't need crm ignore, as they should be able to use the normal Update API to fix any bad state. Just need to check the current state so we don't overwrite another change in progress.